sabnzbd/dirscanner.py

```diff
@@ -115,99 +109,171 @@ def save(self):
+                            if sabnzbd.WIN32:
+                                stat_tuple = os.stat(path)
+                            else:
+                                stat_tuple = entry.stat()
                         except OSError:
                             continue
-                        if compare_stat_tuple(stat_tuple, stat_tuple_tmp):
-                            break
-                        stat_tuple = stat_tuple_tmp
                     else:
-                        # Not stable
+                        self.ignored[path] = 1
+                        yield path, catdir, None
                         continue
 
-                    # Add the NZB's
-                    res, _ = sabnzbd.nzbparser.add_nzbfile(path, catdir=catdir, keep=False)
-                    if res < 0:
-                        # Retry later, for example when we can't read the file
-                        self.suspected[path] = stat_tuple
-                    elif res == 0:
-                        self.error_reported = False
-                    else:
-                        self.ignored[path] = 1
+                    if path in self.suspected:
+                        if not compare_stat_tuple(self.suspected[path], stat_tuple):
+                            # Suspected file attributes have changed
+                            del self.suspected[path]
+
+                    yield path, catdir, stat_tuple
+        except OSError:
+            if not self.error_reported and not catdir:
+                logging.error(T("Cannot read Watched Folder %s"), filesystem.clip_path(folder))
+                self.error_reported = True
+
+    async def when_stable_add_nzbfile(self, path: str, catdir: Optional[str], stat_tuple: os.stat_result):
+        """Try and import the NZB but wait until the attributes are stable for 1 second, but give up after 3 sec"""
+
+        logging.info("Trying to import %s", path)
+
+        # Wait until the attributes are stable for 1 second, but give up after 3 sec
+        # This indicates that the file is fully written to disk
+        for n in range(3):
+            await asyncio.sleep(1.0)
+
+            try:
+                stat_tuple_tmp = os.stat(path)
+            except OSError:
+                continue
+            if compare_stat_tuple(stat_tuple, stat_tuple_tmp):
+                break
+            stat_tuple = stat_tuple_tmp
+        else:
+            # Not stable
+            return
+
+        # Add the NZB's
+        res, _ = sabnzbd.nzbparser.add_nzbfile(path, catdir=catdir, keep=False)
+        if res < 0:
+            # Retry later, for example when we can't read the file
+            self.suspected[path] = stat_tuple
+        elif res == 0:
+            self.error_reported = False
+        else:
+            self.ignored[path] = 1
+
+    def scan(self):
+        """Schedule a scan of the watched folder"""
+        if not self.loop:
+            return
+
+        if not (dirscan_dir := self.dirscan_dir):
```

Comment:
Why do this and not just check `if self.dirscan_dir` and use `self.dirscan_dir` below in the scan-call?

Reply:
I think that change would be fine if `newdir` and `scan` will both be called from the same thread - I think it's the menu bar on Mac I didn't see a way to trigger it on Windows?

I do similar in other places such as `scanner` because we are in a different thread so as I understand it the directory could change between between when we do an `if` and actually use it so I make a copy.
I think it could be removed from scanner but then `scan_async` would have to check it as well.